File: apps-microservices/crawler-monitor-backend/tests/albums.test.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import express from 'express';
import jwt from 'jsonwebtoken';
import request from 'supertest';
import { mountAlbumsRouter } from '../src/lib/albums.js';

const JWT_SECRET = 'test-secret';

// Helper: forge a JWT for an admin user.
function adminToken() {
  return jwt.sign({ role: 'admin' }, JWT_SECRET, { expiresIn: '1h' });
}

// Helper: a minimal authenticateToken middleware mirroring the one in server.js.
function authenticateToken(req, res, next) {
  const authHeader = req.headers['authorization'];
  const token = authHeader && authHeader.split(' ')[1];
  if (!token) return res.sendStatus(401);
  jwt.verify(token, JWT_SECRET, (err, user) => {
    if (err) return res.sendStatus(403);
    req.user = user;
    next();
  });
}

// Stub auditMiddleware so we can record every (action, options) call.
function makeAuditStub() {
  const calls = [];
  const mw = (action, options) => {
    calls.push({ action, options });
    return (req, _res, next) => next();
  };
  return { mw, calls };
}

// fakeFetch builder: lets each test assert on the URL/method/body forwarded
// to the Python service and craft a controlled response.
function makeFakeFetch(impl) {
  const calls = [];
  const fn = async (url, opts) => {
    calls.push({ url, opts });
    return impl(url, opts, calls.length);
  };
  return { fn, calls };
}

// Convenience to wire up a fresh Express app for each test.
function buildApp({ auditMiddleware, fetchFn, baseUrl = 'http://image-download:8505', destructiveLimit }) {
  const app = express();
  app.set('trust proxy', 1);
  app.use(express.json());
  app.use(
    '/api/albums',
    authenticateToken,
    mountAlbumsRouter({ auditMiddleware, fetchFn, baseUrl, destructiveLimit }),
  );
  return app;
}

test('GET / requires JWT (401 without)', async () => {
  const { mw } = makeAuditStub();
  const { fn } = makeFakeFetch(() => ({ status: 200, headers: new Map([['content-type', 'application/json']]), json: async () => ([]) }));
  const app = buildApp({ auditMiddleware: mw, fetchFn: fn });
  const res = await request(app).get('/api/albums/');
  assert.equal(res.status, 401);
});

test('GET / forwards to /domains/_summary and returns JSON', async () => {
  const { mw } = makeAuditStub();
  const { fn, calls } = makeFakeFetch(() => ({
    status: 200,
    headers: new Map([['content-type', 'application/json']]),
    json: async () => ([{ domain: 'example.com', images: 42 }]),
  }));
  const app = buildApp({ auditMiddleware: mw, fetchFn: fn });
  const res = await request(app)
    .get('/api/albums/')
    .set('Authorization', `Bearer ${adminToken()}`);
  assert.equal(res.status, 200);
  assert.equal(calls.length, 1);
  assert.match(calls[0].url, /\/domains\/_summary$/);
  assert.equal(calls[0].opts.method, 'GET');
  assert.deepEqual(res.body, [{ domain: 'example.com', images: 42 }]);
});

test('GET /:domain/products forwards query string to upstream', async () => {
  const { mw } = makeAuditStub();
  const { fn, calls } = makeFakeFetch(() => ({
    status: 200,
    headers: new Map([['content-type', 'application/json']]),
    json: async () => ({ items: [], total: 0 }),
  }));
  const app = buildApp({ auditMiddleware: mw, fetchFn: fn });
  const res = await request(app)
    .get('/api/albums/example.com/products?page=2&limit=50&search=foo')
    .set('Authorization', `Bearer ${adminToken()}`);
  assert.equal(res.status, 200);
  assert.equal(calls.length, 1);
  // Path AND query must be forwarded.
  assert.match(calls[0].url, /\/domains\/example\.com\/products\?/);
  assert.match(calls[0].url, /page=2/);
  assert.match(calls[0].url, /limit=50/);
  assert.match(calls[0].url, /search=foo/);
});

test('DELETE /:domain/products/:id/images/:filename → 204 propagated', async () => {
  const { mw, calls: auditCalls } = makeAuditStub();
  const { fn, calls } = makeFakeFetch(() => ({
    status: 204,
    headers: new Map(),
  }));
  const app = buildApp({ auditMiddleware: mw, fetchFn: fn });
  const res = await request(app)
    .delete('/api/albums/example.com/products/abc123/images/img-001.jpg')
    .set('Authorization', `Bearer ${adminToken()}`);
  assert.equal(res.status, 204);
  assert.equal(calls.length, 1);
  assert.match(calls[0].url, /\/images\/example\.com\/abc123\/img-001\.jpg$/);
  assert.equal(calls[0].opts.method, 'DELETE');
  // auditMiddleware was registered with the right action name on this route.
  assert.ok(auditCalls.some((c) => c.action === 'delete_image'));
});

test('DELETE /:domain → 202 propagated with body { job_id }', async () => {
  const { mw, calls: auditCalls } = makeAuditStub();
  const { fn } = makeFakeFetch(() => ({
    status: 202,
    headers: new Map([['content-type', 'application/json']]),
    json: async () => ({ job_id: 'delete-album-uuid', status: 'queued' }),
  }));
  const app = buildApp({ auditMiddleware: mw, fetchFn: fn });
  const res = await request(app)
    .delete('/api/albums/example.com')
    .set('Authorization', `Bearer ${adminToken()}`);
  assert.equal(res.status, 202);
  assert.deepEqual(res.body, { job_id: 'delete-album-uuid', status: 'queued' });
  assert.ok(auditCalls.some((c) => c.action === 'delete_album'));
});

test('rate-limit: 11th DELETE within window → 429 (limit=10/min)', async () => {
  const { mw } = makeAuditStub();
  const { fn } = makeFakeFetch(() => ({ status: 204, headers: new Map() }));
  const app = buildApp({ auditMiddleware: mw, fetchFn: fn, destructiveLimit: 10 });
  const tok = adminToken();
  // 10 DELETEs should all succeed.
  for (let i = 0; i < 10; i++) {
    const r = await request(app)
      .delete(`/api/albums/example.com/products/p${i}`)
      .set('Authorization', `Bearer ${tok}`);
    assert.equal(r.status, 204, `request #${i + 1} expected 204, got ${r.status}`);
  }
  // 11th must hit the limiter.
  const over = await request(app)
    .delete('/api/albums/example.com/products/p11')
    .set('Authorization', `Bearer ${tok}`);
  assert.equal(over.status, 429);
});

test('rate-limit: GETs are NOT rate-limited even after many calls', async () => {
  const { mw } = makeAuditStub();
  const { fn } = makeFakeFetch(() => ({
    status: 200,
    headers: new Map([['content-type', 'application/json']]),
    json: async () => ([]),
  }));
  const app = buildApp({ auditMiddleware: mw, fetchFn: fn, destructiveLimit: 5 });
  const tok = adminToken();
  for (let i = 0; i < 20; i++) {
    const r = await request(app)
      .get('/api/albums/')
      .set('Authorization', `Bearer ${tok}`);
    assert.equal(r.status, 200, `GET #${i + 1} expected 200, got ${r.status}`);
  }
});

test('auditMiddleware is invoked for every destructive route', async () => {
  const { mw, calls } = makeAuditStub();
  const { fn } = makeFakeFetch(() => ({ status: 204, headers: new Map() }));
  buildApp({ auditMiddleware: mw, fetchFn: fn });
  // Destructive routes registered: sync, redownload x2, delete x3 = 6 audited routes.
  // Read routes (GET) must NOT be audited.
  const actions = calls.map((c) => c.action);
  assert.ok(actions.includes('sync_album'), `expected sync_album, got ${actions}`);
  assert.ok(actions.includes('redownload_product'), `expected redownload_product, got ${actions}`);
  assert.ok(actions.includes('redownload_image'), `expected redownload_image, got ${actions}`);
  assert.ok(actions.includes('delete_album'), `expected delete_album, got ${actions}`);
  assert.ok(actions.includes('delete_product'), `expected delete_product, got ${actions}`);
  assert.ok(actions.includes('delete_image'), `expected delete_image, got ${actions}`);
  // Make sure no audit was registered for GET-only actions.
  assert.ok(!actions.some((a) => a.startsWith('get_') || a === 'list_albums'));
});

test('POST /:domain/sync forwards body and returns upstream response', async () => {
  const { mw } = makeAuditStub();
  const { fn, calls } = makeFakeFetch(() => ({
    status: 202,
    headers: new Map([['content-type', 'application/json']]),
    json: async () => ({ job_id: 'sync-uuid' }),
  }));
  const app = buildApp({ auditMiddleware: mw, fetchFn: fn });
  const res = await request(app)
    .post('/api/albums/example.com/sync')
    .set('Authorization', `Bearer ${adminToken()}`)
    .send({ force: true });
  assert.equal(res.status, 202);
  assert.match(calls[0].url, /\/sync\/example\.com$/);
  assert.equal(calls[0].opts.method, 'POST');
  // Body forwarded as JSON
  assert.equal(calls[0].opts.headers['content-type'], 'application/json');
  assert.equal(JSON.parse(calls[0].opts.body).force, true);
});

File: apps-microservices/crawler-monitor-backend/tests/albums_test.go
package tests

import (
	"bytes"
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync"
	"sync/atomic"
	"testing"

	"github.com/Hellopro-fr/crawler-monitor-backend/internal/config"
	"github.com/Hellopro-fr/crawler-monitor-backend/internal/httpapi"
	"github.com/Hellopro-fr/crawler-monitor-backend/internal/store/redisstore"
	"github.com/alicebob/miniredis/v2"
)

// recordingAuditAlbums captures all audit Append calls (separate from helpers_test
// recordingAudit to avoid type confusion in mixed test runs).
type recordingAuditAlbums struct {
	mu      sync.Mutex
	Entries []map[string]any
}

func (r *recordingAuditAlbums) Append(_ context.Context, e map[string]any) error {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.Entries = append(r.Entries, e)
	return nil
}

func (r *recordingAuditAlbums) all() []map[string]any {
	r.mu.Lock()
	defer r.mu.Unlock()
	out := make([]map[string]any, len(r.Entries))
	copy(out, r.Entries)
	return out
}

// upstream stub that records every call.
type upstreamSpy struct {
	mu       sync.Mutex
	URLs     []string
	Methods  []string
	Bodies   []string
	respond  func(w http.ResponseWriter, r *http.Request)
	count    atomic.Int64
}

func newUpstreamSpy(respond func(http.ResponseWriter, *http.Request)) (*httptest.Server, *upstreamSpy) {
	spy := &upstreamSpy{respond: respond}
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		spy.count.Add(1)
		body, _ := io.ReadAll(r.Body)
		spy.mu.Lock()
		spy.URLs = append(spy.URLs, r.URL.RequestURI())
		spy.Methods = append(spy.Methods, r.Method)
		spy.Bodies = append(spy.Bodies, string(body))
		spy.mu.Unlock()
		spy.respond(w, r)
	}))
	return srv, spy
}

func setupAlbumsTest(t *testing.T, audit httpapi.AuditAppender, upstream *httptest.Server) (*httptest.Server, string) {
	t.Helper()
	mr, _ := miniredis.Run()
	t.Cleanup(mr.Close)
	rs, _ := redisstore.New("redis://" + mr.Addr())
	t.Setenv("IMAGE_DOWNLOAD_SERVICE_URL", upstream.URL)
	cfg := &config.Config{JWTSecret: "test-secret"}
	srv := httptest.NewServer(httpapi.NewRouter(httpapi.Deps{
		Config: cfg, RedisStore: rs, AuditStore: audit,
	}))
	t.Cleanup(srv.Close)
	return srv, mintToken("admin", "test-secret")
}

func TestAlbums_GET_RequiresJWT(t *testing.T) {
	upstream, _ := newUpstreamSpy(func(w http.ResponseWriter, r *http.Request) { w.WriteHeader(200) })
	defer upstream.Close()
	srv, _ := setupAlbumsTest(t, &recordingAuditAlbums{}, upstream)
	resp, _ := http.Get(srv.URL + "/api/albums/")
	if resp.StatusCode != 401 {
		t.Errorf("status=%d, want 401", resp.StatusCode)
	}
}

func TestAlbums_GET_ForwardsSummary(t *testing.T) {
	upstream, spy := newUpstreamSpy(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`[{"domain":"example.com","images":42}]`))
	})
	defer upstream.Close()
	srv, tok := setupAlbumsTest(t, &recordingAuditAlbums{}, upstream)
	resp, err := authedGet(srv.URL+"/api/albums/", tok)
	if err != nil {
		t.Fatal(err)
	}
	if resp.StatusCode != 200 {
		t.Fatalf("status=%d", resp.StatusCode)
	}
	if len(spy.URLs) != 1 || !strings.HasSuffix(spy.URLs[0], "/domains/_summary") {
		t.Errorf("upstream URL=%v", spy.URLs)
	}
	if spy.Methods[0] != "GET" {
		t.Errorf("method=%s", spy.Methods[0])
	}
	body, _ := io.ReadAll(resp.Body)
	if !strings.Contains(string(body), "example.com") {
		t.Errorf("body=%s", body)
	}
}

func TestAlbums_GET_ForwardsQueryString(t *testing.T) {
	upstream, spy := newUpstreamSpy(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"items":[],"total":0}`))
	})
	defer upstream.Close()
	srv, tok := setupAlbumsTest(t, &recordingAuditAlbums{}, upstream)
	_, err := authedGet(srv.URL+"/api/albums/example.com/products?page=2&limit=50&search=foo", tok)
	if err != nil {
		t.Fatal(err)
	}
	if len(spy.URLs) != 1 {
		t.Fatalf("upstream calls=%d", len(spy.URLs))
	}
	got := spy.URLs[0]
	for _, want := range []string{"/domains/example.com/products", "page=2", "limit=50", "search=foo"} {
		if !strings.Contains(got, want) {
			t.Errorf("URL %q missing %q", got, want)
		}
	}
}

func TestAlbums_DELETE_204Propagated(t *testing.T) {
	upstream, spy := newUpstreamSpy(func(w http.ResponseWriter, r *http.Request) { w.WriteHeader(204) })
	defer upstream.Close()
	audit := &recordingAuditAlbums{}
	srv, tok := setupAlbumsTest(t, audit, upstream)

	req, _ := http.NewRequest("DELETE", srv.URL+"/api/albums/example.com/products/abc123/images/img-001.jpg", nil)
	req.Header.Set("Authorization", "Bearer "+tok)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		t.Fatal(err)
	}
	if resp.StatusCode != 204 {
		t.Errorf("status=%d, want 204", resp.StatusCode)
	}
	if len(spy.URLs) != 1 || !strings.HasSuffix(spy.URLs[0], "/images/example.com/abc123/img-001.jpg") {
		t.Errorf("upstream URL=%v", spy.URLs)
	}
	if spy.Methods[0] != "DELETE" {
		t.Errorf("method=%s", spy.Methods[0])
	}
	entries := audit.all()
	if len(entries) != 1 || entries[0]["action"] != "delete_image" {
		t.Errorf("audit entries=%v", entries)
	}
}

func TestAlbums_POST_ForwardsBody(t *testing.T) {
	upstream, spy := newUpstreamSpy(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(202)
		_, _ = w.Write([]byte(`{"job_id":"sync-uuid"}`))
	})
	defer upstream.Close()
	audit := &recordingAuditAlbums{}
	srv, tok := setupAlbumsTest(t, audit, upstream)

	req, _ := http.NewRequest("POST", srv.URL+"/api/albums/example.com/sync", bytes.NewBufferString(`{"force":true}`))
	req.Header.Set("Authorization", "Bearer "+tok)
	req.Header.Set("Content-Type", "application/json")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		t.Fatal(err)
	}
	if resp.StatusCode != 202 {
		t.Errorf("status=%d", resp.StatusCode)
	}
	var body map[string]any
	_ = json.NewDecoder(resp.Body).Decode(&body)
	if body["job_id"] != "sync-uuid" {
		t.Errorf("body=%v", body)
	}
	if len(spy.URLs) != 1 || !strings.HasSuffix(spy.URLs[0], "/sync/example.com") {
		t.Errorf("upstream URL=%v", spy.URLs)
	}
	if spy.Bodies[0] != `{"force":true}` {
		t.Errorf("body forwarded=%s", spy.Bodies[0])
	}
	if !auditHasAction(audit.all(), "sync_album") {
		t.Error("audit action sync_album missing")
	}
}

func TestAlbums_RateLimit_11thDeleteRejected(t *testing.T) {
	upstream, _ := newUpstreamSpy(func(w http.ResponseWriter, r *http.Request) { w.WriteHeader(204) })
	defer upstream.Close()
	srv, tok := setupAlbumsTest(t, &recordingAuditAlbums{}, upstream)

	for i := 0; i < 10; i++ {
		req, _ := http.NewRequest("DELETE", srv.URL+"/api/albums/example.com/products/p"+strconvI(i), nil)
		req.Header.Set("Authorization", "Bearer "+tok)
		resp, err := http.DefaultClient.Do(req)
		if err != nil {
			t.Fatal(err)
		}
		if resp.StatusCode != 204 {
			t.Errorf("call #%d: status=%d", i, resp.StatusCode)
		}
		resp.Body.Close()
	}
	req, _ := http.NewRequest("DELETE", srv.URL+"/api/albums/example.com/products/p11", nil)
	req.Header.Set("Authorization", "Bearer "+tok)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		t.Fatal(err)
	}
	if resp.StatusCode != 429 {
		t.Errorf("11th call: status=%d, want 429", resp.StatusCode)
	}
}

func TestAlbums_GETs_NotRateLimited(t *testing.T) {
	upstream, _ := newUpstreamSpy(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`[]`))
	})
	defer upstream.Close()
	srv, tok := setupAlbumsTest(t, &recordingAuditAlbums{}, upstream)

	for i := 0; i < 20; i++ {
		resp, err := authedGet(srv.URL+"/api/albums/", tok)
		if err != nil {
			t.Fatal(err)
		}
		if resp.StatusCode != 200 {
			t.Errorf("GET #%d: status=%d", i+1, resp.StatusCode)
		}
		resp.Body.Close()
	}
}

func TestAlbums_DestructiveAuditCoverage(t *testing.T) {
	upstream, _ := newUpstreamSpy(func(w http.ResponseWriter, r *http.Request) {
		// Some routes return 202, others 204. Either is fine for audit purposes.
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(202)
		_, _ = w.Write([]byte(`{}`))
	})
	defer upstream.Close()
	audit := &recordingAuditAlbums{}
	srv, tok := setupAlbumsTest(t, audit, upstream)

	calls := []struct {
		method, path, want string
	}{
		{"POST", "/api/albums/d.com/sync", "sync_album"},
		{"POST", "/api/albums/d.com/products/p1/redownload", "redownload_product"},
		{"POST", "/api/albums/d.com/products/p1/images/i1.jpg/redownload", "redownload_image"},
		{"DELETE", "/api/albums/d.com", "delete_album"},
		{"DELETE", "/api/albums/d.com/products/p1", "delete_product"},
		{"DELETE", "/api/albums/d.com/products/p1/images/i1.jpg", "delete_image"},
	}
	for _, c := range calls {
		req, _ := http.NewRequest(c.method, srv.URL+c.path, nil)
		req.Header.Set("Authorization", "Bearer "+tok)
		resp, err := http.DefaultClient.Do(req)
		if err != nil {
			t.Fatal(err)
		}
		resp.Body.Close()
	}
	for _, c := range calls {
		if !auditHasAction(audit.all(), c.want) {
			t.Errorf("missing audit action %q", c.want)
		}
	}
}

func auditHasAction(entries []map[string]any, action string) bool {
	for _, e := range entries {
		if e["action"] == action {
			return true
		}
	}
	return false
}

func strconvI(i int) string {
	if i < 10 {
		return string(rune('0' + i))
	}
	return string(rune('0'+i/10)) + string(rune('0'+i%10))
}
